Add interval marks

bkutil opened the pull request:
Aggregated points coming from `reduce -every` or `batch` operations
represent a quantity over a certain time interval.

This PR modifies `reduce_every` and `reduce_batch` processors so they
include this info into the points as an `interval` field.

This allows views to display these points differently, e.g. by including
the time interval.

Batch and reduce also insert the interval info into marks, so they
can be consumed by reduces further in the flowgraph - e.g. when a `batch`
is followed by several `reduce first()` processors.

Other processors simply forward the marks, including the interval info,
with the notable exception of join.

Commit:
juttle-test-utils.js - test view handles properties in marks

Marks are now represented by an object. Modify the test view so it forwards
all properties of a mark if `view -times true -marks true` options are
specified.

Commit:
batch.js - batch proc inserts intervals into marks

Batch inserts interval information into marks, so that `reduce_batch`
further in the flowgraph can add intervals to reduced points.

bkutil commented:
cc: @dmajda @demmer 

refs: #123

dmajda commented:
What is the plan for `join`?

Overall, I think it would help to split changes needed to add intervals to marks form the ones needed to add intervals to points.

bkutil commented:
The join logic is somewhat complicated, so the plan is:

1. figure out where and why join emits marks
2. figure out the correct `interval` value, if there is one.

Maybe Will would be able to help here.

We can track this as a separate issue, to unblock further work on handling interval info in the viz views.

bkutil commented:
Split the changes between points/marks updates, renamed `mark_interval` to `interval` and omitted the incorrect addition of `interval` info into `witness_timestamp`.

dmajda commented:
Looking at the change now, it seems like the `interval` parameter in `first_epoch` and `other_epoch` isn’t really necessary as in the periodic case the value is always `this.interval`. I’d remove it unless you have some deeper reason to keep it.

> omitted the incorrect addition of `interval` info in `witness_timestamp`

I think the right thing to do here is actually to add an interval with a duration between the last emitted mark and this one. Basically, we are emitting a shortened batch. Note we can bump into the same problems with calendarness as in fanin.

> We can track this as a separate issue, to unblock further work on handling interval info in the viz views.

Well, I think we want to get into the state where all marks (except the very first one) have a correct interval — only then will points coming from a batched reduce always have a correct interval. I’d avoid knowingly pushing any code which can potentially generate incorrect intervals. 

AFAIK, to get into that target state we need to:

 * resolve the `witness_timestamp` mark problem above
 * implement correct semantics for merging streams in fanin
 * fix `join`

I think this should be tracked together, either in #123 (because without this the logic won’t work in all cases) or separately.

bkutil commented:
> implement correct semantics for merging streams in fanin

Lets continue the discussion from https://github.com/juttle/juttle/issues/523#issuecomment-195288571 here:

> output stream of fanin contains marks from all input streams, combined. This creates batches of various sizes in the output stream
>> As for 2) (, the result of discussion with @demmer is that batches of varying length are OK, but we’ll forbid merging streams containing calendar-based batches with streams containing non-calendar-based ones. This will prevent any calendarness loss in batch interval lengths.

>> If this semantics proves to be too restrictive in practice, we can safely revert this position, allow the mixing, and accept possible calendarness loss.

(Calendarness loss means combining streams with calendar (i.e. interval where `interval.duration.months != 0`) and non-calendar intervals. E.g. `:month:` and `:5 days:`)

There are two additional things to consider:

1. In addition to the scenario above, calendarness loss can also occur if `-on` option is used with non-aligned calendar intervals. For example: `reduce -every :month: -on :day 10: ...; reduce -every :month: -on :day 20: | view...` can result in interval of days, which are not calendar. So we'd have to disallow this case as well.

2. It seems we'd have to distinguish batched & reduced streams from batched & non-reduced:

   * In streams coming out of a reduce, both marks and points contain the same interval value. When the interval fragments after `fanin`, the proc recomputes the interval for an incoming mark. However, now the interval may no longer correspond with the value in the reduced point. In this case, it seems that marks should be dropped - `fanin` would emit unbatched stream of reduced points.
  * If the input stream contains only non-reduced points and marks, `fanin` should merge incoming batches in the way we discussed, so the points can be reduced in those (possibly smaller/non-uniform) batches.
  
   As @dmajda pointed out, this approach creates two semantically different modes in the flow,but basically reflects the fact that we now have 'points in time' and 'points as intervals'.

@demmer thoughts?

bkutil committed:
join.js - add interval info to marks

Join accepts batched streams, but only uses marks from input to set the
time of points in groups to mark.time+-epsilon and to flag the input
as batched. The marks themselves are dropped before joining.

Only after the joined points are emitted, if the appropriate inputs were
batched (see batched_leader() for the logic), it restores the
surrounding marks again. In this case, the interval boundaries
will be start_time and complete_time of the joined group. Marks
will contain the interval info.

dmajda commented:
We discussed the problems with @demmer. The results are:

  * We will just allow calendarness loss for marks in situations when it is unavoidable rather than forbid these situations. At the same time we will guarantee that fanin doesn’t lose calendarness at least in the following cases:

    * One input stream with calendar batches and zero or more unbatched streams.
    * One or more input stream with exactly aligned calendar batches.

    This work is covered by #123 and this PR (which will be updated to make sure all procs handle marks with intervals correctly and that fanin behaves as described above).

  * We won’t allow points with intervals to be “split” by marks. This is covered by #617.
  * We will make the `interval` field special. This is covered by #618.

bkutil committed:
fanin.js - handle intervals in marks

Fanin merges marks from multiple streams and re-calculates the
intervals, while trying to preserve their 'calendarness'.

- If there are multiple marks with the same timestamp, but with
  different intervals, pick the one with the shortest interval.

  This ensures that if there is a period after which the mark times
  overlap again, the same interval will be used consistently in the output
  stream. Also, if all inputs have calendar marks, the property is
  preserved, as there is no need to subtract the shorter interval from the
  longer one.

  E.g. for :1M: and :2M: batches, the interval will always be :1M:.

  If a calendar and a non-calendar stream is mixed, calendarness is lost
  in mark intervals.

- If streams with non-overlapping marks are merged, the interval is set
  to be the distance since the last mark (first mark doesn't have
  interval set for multiple stream fanin).

bkutil committed:
fanin.js - move fanin earliest-is-eof case to the main if block

This makes the 'emit earliest' loop a bit more readable, as all
point types are handled in a single if/elseif/else statement.

bkutil committed:
pace.js - forward marks

Pace proc forwards all information received in marks, not just time.

When the marks will contain interval information (in upcoming commits,
this will ensure it is sent out as well.

bkutil commented:
The fanin now behaves as described. Also updated pace to forward marks, which I missed.

bkutil commented:
@dmajda can I have another review, please?

dmajda commented:
The commit message of 79353f5665b3ee594c6dd2f94f61ccee24872da8 (“periodic-fanin.js, reduce.js - emit marks with interval information”) no longer corresponds to changes in the commit, which now only contains changes that prepare for passing marks with intervals around.

It would also be good to move this commit before adapting `batch`.

dmajda commented on lib/runtime/procs/reduce.js:
```diff
@@ -152,8 +152,16 @@ class reduce_base extends periodic_fanin {
+                    pt.time = eob;
+                }
+                // ...and interval info from marks coming from batch,
+                // every of `reduce -every`, or undefined for unbatched
```
The comment is slightly misleading — in case of unbatched stream, the interval isn’t set to `undefined` but not set at all. This is especially confusing because `undefined` is not a valid Juttle value.

dmajda commented:
Please squash f8b8e6949034bc2e1d11dd6d132d605b307f697c to 79353f5665b3ee594c6dd2f94f61ccee24872da8.

dmajda commented:
I also think both `reduce` changes can just go in one commit. That is, squash f081b6ad02a76ae2a7bccd0ff77f249d37f2bd52 into c8d9c74107994cfcdfd2a2059b5bf11161779b5f.

dmajda commented on lib/runtime/procs/join.js:
```diff
@@ -801,28 +801,44 @@ class join extends base {
+        var interval;
+
+        if (leader) {
+            interval = leader.live().complete_time.subtract(leader.live().start_time);
```
This means we will lose calendarness even in cases mentioned in my [earlier comment](https://github.com/juttle/juttle/pull/581#issuecomment-197864965). Can we avoid that?

bkutil replied:
Thats a good question. I think it might be possible to copy information from mark into the join groups and then extract them after joining.

bkutil replied:
Ok, that worked. Added it into `join` code and added a few tests.

dmajda commented on lib/runtime/procs/join.js:
```diff
@@ -801,28 +801,44 @@ class join extends base {
         if (points.length) {
+            // Marks around points, add interval info
+            if (interval) {
+                points.map(function(pt) { pt.interval = interval; });
```
I don’t think it makes sense to add time to points in `join`. Unlike in `reduce`, they are not synthetic and they don’t cover the whole interval.

dmajda commented on lib/runtime/procs/fanin.js:
```diff
@@ -130,11 +131,22 @@ class Input {
 
+        function hasInterval(a) {
+            return a.mark && a.interval;
+        }
```
The name of the function is misleading — it checks more than it says. Either rename to something like `isMarkWithInterval` or just inline it.

dmajda commented on lib/runtime/procs/fanin.js:
```diff
@@ -130,11 +131,22 @@ class Input {
             var earliestPt = earliest.first();
 
             if (inPt.time.lt(earliestPt.time) || (inPt.mark && inPt.time.eq(earliestPt.time))) {
+                if (hasInterval(inPt) && hasInterval(earliestPt)) {
```
It would be good to add a comment explaining why is this done and a test verifying the behavior.

dmajda commented on lib/runtime/procs/pace.js:
```diff
@@ -116,15 +116,17 @@ class pace extends fanin {
         }
     }
     mark(mark) {
+        mark.time = this.offset_time(mark.time);
```
You should make a copy before modifying in order to avoid influencing other procs working with the mark (e.g. in a parallel stream).

dmajda commented:
Squash 54ac58e0c933ae1ee40a147e8db58e4b5b5a4d9c into 223294d0054ac2b26dbb5d946077784254a1fbdb? That would also resolve part of my [comment](https://github.com/juttle/juttle/pull/581/commits/223294d0054ac2b26dbb5d946077784254a1fbdb#r56816565) there.

dmajda commented:
Beside the squashing, it would be nice to regroup the PR so that changes that enable sending marks with intervals around go in first, followed by changes causing generating marks with intervals, followed by the change that causes generating points with intervals in `reduce`. That would follow the sequence of what is necessary for what, which would be easier to follow compared to current somewhat zig-zaggy approach.

bkutil committed:
Fanin, reduce and periodic fanin forward marks

When emitting or re-emitting mark, do not send only time, but forward
complete mark.

When the marks will contain interval info (in the upcoming commits),
this ensures it is sent out as well.

bkutil committed:
reduce.js - insert interval info into aggregated points

Aggregated points coming from `batch` operation represent a quantity over a
certain time interval.

This patch modifies `reduce_every` and `reduce_batch` processors so they
include this info into the points as an `interval` field.

This allows views to display these points differently, e.g. by including
the time interval.

bkutil commented:
@dmajda updated according to review suggestions

dmajda commented on lib/runtime/procs/pace.js:
```diff
@@ -116,15 +116,17 @@ class pace extends fanin {
         }
     }
     mark(mark) {
+        mark.time = this.offset_time(mark.time.clone());
```
I meant to clone the mark, not the time.

bkutil replied:
Oops, my bad. Will fix.

dmajda commented on lib/runtime/procs/join.js:
```diff
@@ -801,11 +802,20 @@ class join extends base {
-            this.maybe_emit_mark(leader.live().start_time);
+            var live_leader = leader.live();
+            interval = live_leader.interval;
+            interval = interval || live_leader.complete_time.subtract(live_leader.start_time);
```
When exactly does the situation that we don’t have `live_leader.interval` set arise?

bkutil replied:
When the mark in a live batched input does not contain an interval. This could happen for the first mark (group). But it should be possible to avoid this case by taking the interval from the 'end-of-batch' mark of the first batch, instead of the initial one.

dmajda commented on test/runtime/specs/juttle-spec/juttle-procs-join.spec.md:
```diff
@@ -1079,3 +1079,65 @@ marked complete and participates in early joins. Later joins get the second tabl
+
+### Output
+
+    { "time": "1970-01-01T00:00:00.000Z", "interval": "1M", "mark": true }
```
This is odd, shouldn’t the first mark be interval-less?

bkutil replied:
Ok, if we want to do this, we should be consistent across procs. I'll modify batch, reduce, fanin and join to emit the first mark without interval info. I'll put it into a separate commit, so it can be reverted later, if we decide against it.

dmajda commented on test/runtime/specs/juttle-spec/juttle-procs-batch.spec.md:
```diff
@@ -180,12 +180,28 @@ no bare numbers with the new -every option
 
 ### Output
-    { "time": "1970-01-01T00:00:02.000Z", "mark": true }
+    { "time": "1970-01-01T00:00:02.000Z", "interval": "00:00:02.000", "mark": true }
```
Same here — I think the first mark should be interval-less.

bkutil committed:
First mark should not carry interval information

Marks carry information about the preceding batch. For the first mark,
there is no such batch and thus, the interval should not be defined.

Modify join, fanin, batch and reduce to keep track of emitted marks
and do not add/omit interval information from the first one.
